### PR TITLE
Enhancement/use hpp for headers

### PR DESCRIPTION
**Pull request by TSNoble**
### Description

First of all, apologies for the huge PR 😅

- Partially addresses #857 by replacing all .h files within moveit with .hpp files and, where possible, modifying imports to use .hpp
- Adds a script to autogenerate `.h` files from `.hpp` files with a deprecation warning (see comments)
- Updates MIGRATION.md

The main aim of this PR is to tackle a good portion of the low-hanging fruit so that subsequent (and less intrusive) PRs can be raised for specific types of imports.

I've left alone the following types of imports:
  - Imports to C libraries e.g. <math.h> (will raise a PR using C++ equivalents)
  - Imports to <ros/ros.h> (will raise a PR using rclcpp imports)
  - Imports to ROS1 msg packages (will raise a PR using ROS2 equivalents where possible)
  - Imports to <fmtlib/fmt.h> (I might raise a PR replacing these with \<format\> if there aren't downsides)  
  - Imports to ROS2 dependencies where I'm unsure if .hpp files exist e.g. <tf2/etc.h>
  - Imports non-ROS libraries e.g. <kdl/etc.h> as I'm unsure if .hpp files exist (possibly worth tackling one dependency at a time)
  - Imports of the form <moveit_..._export.h> (I couldn't get these working, but would be happy to raise a PR with guidance)
    - It seems they come from the GENERATE_EXPORT_HEADER() CMake macro, which generates .h by default.
    - I wonder if we could get it to produce .hpp files?

The import style seems quite inconsistent, I've seen "", <>, and relative imports used for internal headers
  - It might be worth raising an issue to standardise this, if there isn't one already
  - It might be nice to enforce some kind of ordering (I'm biased towards built-in, external, internal, separated by newlines)

**I'm a bit concerned about breaking downstream projects which import these headers.**

I wonder if it'd be useful to add a script to the CI that generates a .h for every .hpp, which simply imports the .hpp and prints a deprecation warning?

Something like the following for all files "x.h":

```
#warning .h imports are deprecated. Consider using .hpp
#include <x.hpp>
```

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/) (Doesn't affect user-facing functionality)
- [x] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/moveit/moveit2/blob/main/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html) (existing tests passing should be a good signifier)
- [x] Include a screenshot if changing a GUI (not changing GUI per-se, just the imports used within)
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/moveit/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by TSNoble** [comment]
Something like the following might work for backwards-compatibility:

```py
import sys
from pathlib import Path

ALL_HPPS = "**/*.hpp"
USAGE = "Usage: ./generate_deprecated_headers <install_directory>"
DEPRECATION = "{0}.h imports are deprecated. Consider using {0}.hpp"

def generate_c_header(hpp_path: str) -> None:
    header_directory = hpp_path.parent
    file = hpp_path.name.replace(".hpp", "")
    with open(header_directory/f"{file}.h", "w") as h_file:
        h_file.write(f"#warning {DEPRECATION.format(file)}\n")
        h_file.write(f"#include <{file}.hpp>\n")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        sys.exit(USAGE)
    install_directory = sys.argv[1]
    for hpp_path in Path(install_directory).rglob(ALL_HPPS):
        generate_c_header(hpp_path)
```

This could be stored under `moveit/scripts` (and should probably contain a list of moveit packages to avoid wreaking havoc on the entire install space 😅 )

**Comment by TSNoble** [comment]
This seems to work as expected:

```sh
root@66ba015015ae:~/ws_moveit# ./src/moveit2/moveit/scripts/create_deprecated_headers.py install/
root@66ba015015ae:~/ws_moveit# find ./install/moveit_core/ -type f -name '*.h' | head -10
./install/moveit_core/include/moveit_butterworth_filter_parameters/moveit_butterworth_filter_parameters.h
./install/moveit_core/include/moveit_ruckig_filter_parameters/moveit_ruckig_filter_parameters.h
./install/moveit_core/include/moveit_acceleration_filter_parameters/moveit_acceleration_filter_parameters.h
./install/moveit_core/include/moveit_core/moveit_collision_detection_fcl_export.h
./install/moveit_core/include/moveit_core/moveit/controller_manager/controller_manager.h
./install/moveit_core/include/moveit_core/moveit/robot_state/robot_state.h
./install/moveit_core/include/moveit_core/moveit/robot_state/conversions.h
./install/moveit_core/include/moveit_core/moveit/robot_state/attached_body.h
./install/moveit_core/include/moveit_core/moveit/robot_state/cartesian_interpolator.h
./install/moveit_core/include/moveit_core/moveit/constraint_samplers/constraint_sampler_allocator.h
root@66ba015015ae:~/ws_moveit# cat ./install/moveit_core/include/moveit_butterworth_filter_parameters/moveit_butterworth_filter_parameters.h
#warning moveit_butterworth_filter_parameters.h imports are deprecated. Consider using moveit_butterworth_filter_parameters.hpp
#include <moveit_butterworth_filter_parameters.hpp>
```

**Comment by TSNoble** [comment]
Compared with [a passing run](https://github.com/moveit/moveit2/actions/runs/11859496196/job/33052574762?pr=3088):
```
Running moveit_kinematics/ikfast_kinematics_plugin/scripts/create_ikfast_moveit_plugin.py "fanuc" "manipulator" "/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin" "base_link" "tool0" "/tmp/ikfast.jiohdz/.openrave/kinematics.0e9021dc6a6dac745a815e8917dad8a1/ikfast0x10000049.Transform6D.0_1_2_3_4_5.cpp"
Failed to import ament_index_python. No ROS2 environment available? Trying without.
Creating IKFastKinematicsPlugin with parameters: 
 robot_name:           fanuc
 base_link_name:       base_link
 eef_link_name:        tool0
 planning_group_name:  manipulator
 ikfast_plugin_pkg:    /home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin
 ikfast_output_path:   /tmp/ikfast.jiohdz/.openrave/kinematics.0e9021dc6a6dac745a815e[89](https://github.com/moveit/moveit2/actions/runs/11859496196/job/33052574762?pr=3088#step:10:90)17dad8a1/ikfast0x10000049.Transform6D.0_1_2_3_4_5.cpp
 search_mode:          OPTIMIZE_MAX_JOINT
 srdf_filename:        fanuc.srdf
 robot_name_in_srdf:   fanuc
 moveit_config_pkg:    fanuc_moveit_config

Found source code generated by IKFast version 73
Failed to find package: /home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin. Will create it in /home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin.
Created package.xml at: '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/package.xml'
Create parameters.yaml at: '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/src/ikfast_kinematics_parameters.yaml'
Created ikfast header file at '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/include/ikfast.h'
Created ikfast plugin file at '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/src/fanuc_manipulator_ikfast_moveit_plugin.cpp'
Created plugin definition at  '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/fanuc_manipulator_moveit_ikfast_plugin_description.xml'
Created cmake file at '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/CMakeLists.txt'
Wrote package.xml at  '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/package.xml'
Created update plugin script at '/home/runner/work/moveit2/moveit2/fanuc_ikfast_plugin/update_ikfast_plugin.sh'
Failed to update MoveIt package:
Failed to find package: fanuc_moveit_config
```

It seems to encounter the same failure, but continues?

**Comment by TSNoble** [comment]
Ignore the above. I just happened to break the IKFast generator 😅

**Comment by TSNoble** [comment]
@rhaschke I'm a bit confused by the IK failures. I changed the ikfast.hpp template to a hpp and changed the generation script and imports to use hpp.

This caused everything besides one job to fail. I think the failure was due to the job injecting the function robot into the container, which uses a .h include.

I've reversed the template, script, and imports back to ikfast.h, but now everything is failing trying to import this file.

**Comment by rr-tom-noble** [comment]
Also regarding the autogenerated .h files, the current script might be fine. I think it depends on whether we can run it over the local install space (in which case we can simply the filter to all .hpps), or whether we expect it to be integrated in an install space with other packages (after building in the CI)

**Comment by rr-tom-noble** [comment]
> This caused everything besides one job to fail. I think the failure was due to the job injecting faunc robot code into the container, which uses an ikfast.h include.

Correction, not everything but most: See [194d544](https://github.com/moveit/moveit2/pull/3113/commits/194d544121e3349ca266409ea10407f0ecb580d6) runs

**Comment by TSNoble** [comment]
@rhaschke @tylerjw @sea-bass @sjahr 

Unsure who to tag here, but it'd probably be good to get a few set of eyes over this PR. We're already seeing downstream failures from importing ".h" in other packages pulled in by the tutorial images. Does the proposal to autogenerate deprecated ".h" files seem sensible?

**Comment by mikeferguson** [comment]
I'm personally of the opinion that the time to rename files was during the initial port to ROS 2 / MoveIt 2 - and maybe we should save all these headaches for MoveIt 3... I actually have no idea just how much downstream code is going to be impacted by these changes.

**Comment by TSNoble** [comment]
> I'm personally of the opinion that the time to rename files was during the initial port to ROS 2 / MoveIt 2 - and maybe we should save all these headaches for MoveIt 3... I actually have no idea just how much downstream code is going to be impacted by these changes.

With the script proposed, hopefully none. It'll autogenerate c headers from the hips which just import the hpp

**Comment by sjahr** [comment]
@TSNoble Thanks a lot for tackling this!

I'd in favor of merging this PR, even it is a huge breaking change. As you say @mikeferguson it should have been done years ago :sweat_smile: and nobody knows when MoveIt 3 might happen. Let's discuss this further in next weeks maintainer meeting and decide there.

> The import style seems quite inconsistent, I've seen "", <>, and relative imports used for internal headers
> It might be worth raising an issue to standardise this, if there isn't one already
> It might be nice to enforce some kind of ordering (I'm biased towards built-in, external, internal, separated by newlines)

Yes! That's a good idea. Are you aware of any CI that might help us enforcing that? The only thing in this direction I've seen so far `IncludeBlocks: Preserve` in the clang-format file. The blocking structure itself needs to be enforced by reviewers

>I wonder if it'd be useful to add a script to the CI that generates a .h for every .hpp, which simply imports the .hpp and prints a deprecation warning?

 :+1:

**Comment by rr-tom-noble** [comment]
@sjahr 

> I'd in favor of merging this PR, even it is a huge breaking change. As you say @mikeferguson it should have been done years ago 😅 and nobody knows when MoveIt 3 might happen. Let's discuss this further in next weeks maintainer meeting and decide there.

Thanks and yeah, I fully agree that such a large change warrants broader discussion. The maintainer meeting has come at a convenient time! :slightly_smiling_face: 

> Yes! That's a good idea. Are you aware of any CI that might help us enforcing that? The only thing in this direction I've seen so far IncludeBlocks: Preserve in the clang-format file. The blocking structure itself needs to be enforced by reviewers

Unfortunately not for C++. I suppose clang-tidy / clang-format might come with some options?

It might also be useful to add a check to the CI which makes sure there are no .h files prior to running the auto-generator, as a way of maintaining this standard. This isn't possible currently (see the section on imports I didn't touch), but should be once the rest of those are done (which I've now added as tasks on #857)

**Comment by sea-bass** [comment]
To give some precedent; the `urdf` library recently did this on rolling in https://github.com/ros2/urdf/pull/39 -- but they actually still include the old `.h` file in the repo itself under source control with a deprecation warning.

Even with doing that, several downstream repos, including this one, were affected. Because we treat all compiler warnings as errors in CI, we caught the breakage early on. But you can imagine similar effects happening downstream with something like this, just a lot more files.

I would recommend keeping and using the script, but also including the `.h` files as part of this PR instead of having the build system autogenerate them. It's less risky, especially if people are going to be source building MoveIt (a very common workflow, perhaps even more so than binaries to be honest).

Then, maybe by the next ROS release (May 2025), we can see how it would look to remove the `.h` files altogether.

**Comment by TSNoble** [comment]
@sea-bass 

> Even with doing that, several downstream repos, including this one, were affected. Because we treat all compiler warnings as errors in CI, we caught the breakage early on. But you can imagine similar effects happening downstream with something like this, just a lot more files.

I'd argue that this isn't necessarily a bad thing, as it promotes adopting the same good practices downstream. I feel like treating warnings as errors implies that a build _should_ break when good practices aren't being followed.

> I would recommend keeping and using the script, but also including the .h files as part of this PR instead of having the build system autogenerate them. It's less risky, especially if people are going to be source building MoveIt (a very common workflow, perhaps even more so than binaries to be honest).

My main concern is that committing files which can be autogenerated feels like a bad practice, and adds more steps to a developer's workflow. It may be possible to add the auto-generation as part of CMake (see [here](https://stackoverflow.com/questions/49053544/how-do-i-run-a-python-script-every-time-in-a-cmake-build)). Would you be happy with this approach?

**Comment by TSNoble** [comment]
The CMake seems to be working insofar as generating the files, but I think the .hpp imports are a bit messed up.

I'm currently using an import relative to the workspace, so .e.g. an import to version.h would turn into <moveit_core/moveit/version.hpp>.

I've noticed lots of packages skip the package name e.g. <moveit/version.hpp> 

Is any import that starts from the top-level include space valid e.g. <a/b/c/d.hpp> == <b/c/d.hpp> == <c/d.hpp>, or do imports need to start from the moveit dir within a package's install space?

**Comment by TSNoble** [comment]
Also I definitely think an automated test here would be useful for testing all imports.

Something parameterised by each .h which just imports it and checks there are no errors.

(On second though, since the errors would be generated at compile time, I suppose this doesn't even need to be a test. Just an autogenerated set of cpps that we build to test)

Any guidance on where aomething like this should live would be much appreciated.

Something like:
  - Scan for all hpps
  - Generate cpp that just has <include etc.h>
  - Add cpp to package
  - Build cpp

Maybe instead of doing this all in one go, this could be added as extra steps in the `create_deprecated_headers()` macro

**Comment by TSNoble** [comment]
@sea-bass 

> I would recommend keeping and using the script, but also including the .h files as part of this PR instead of having the build system autogenerate them.

I'm having some trouble getting the autgen script to work across all files, so I may take you up on this option. I think I also disagree with my earlier point about it being more developer overhead, since new .hpps implies new functionality (i.e. non-breaking since it won't have been used in other repos yet). A find and delete on all .hs when we're ready to remove them doesn't seem like that much effort.

My current plan is to get the PR fully passing before the maintainer meeting, and we can discuss in-depth there.

**Comment by sea-bass** [comment]
Sounds good. Thanks for your efforts.

BTW the tutorial images build fails because of external repos needing to use their new .hpps. See for example this from the `pick_ik` repo: https://github.com/moveit/moveit2/actions/runs/11921408926/job/33225458963?pr=3113#step:9:44

This is what I meant by the "compiler warnings as errors" alerting people in CI when these changes get in.

**Comment by TSNoble** [comment]
@sjar @sea-bass 

I tweaked the generation script and stripped out the cmake, and used it as the basis for the committed .h files.

It now runs over the source directory before build, as I feel a lot more comfortable determining relative imports from there rather than the install space after build. 

It's based on the following rules:

- Run over all hpp files
- Extract the BSD-3 and  Authorship by stripping out everything after `#pragma once`. This is a bit flaky, and I've added `#pragma once` to a few hpps missing them. I don't _think_ there's a downside to this?
- Add a deprecated message (now using `#pragma message` rather than `#warning`)
- Determine the import by getting the path of the hpp relative to the include directory parent

I've left it in the PR, but as a tool to be manually run as more of a starting point so the developer can focus on some of the more troublesome files. Maybe with some tweaking can be made more robust,

Output of the script:
```
root@4fc39a04273a:~/ws_moveit/src/moveit2# ./moveit/scripts/create_deprecated_headers.py 
Generating deprecated .h files for 414 .hpp files...
Summary: Can generate 376 .h files
Cannot generate .h files for the following files:
/root/ws_moveit/src/moveit2/moveit_planners/pilz_industrial_motion_planner/test/test_utils.hpp
/root/ws_moveit/src/moveit2/moveit_planners/ompl/ompl_interface/test/load_test_robot.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/controller_manager/controller_manager.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/robot_state/robot_state.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/transforms/transforms.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/collision_detection/collision_common.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/collision_detection/world.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/collision_detection/collision_matrix.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/planning_interface/planning_response.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/kinematic_constraints/utils.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/robot_model/joint_model.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/robot_model/joint_model_group.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/robot_model/robot_model.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/robot_trajectory/robot_trajectory.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_core/planning_scene/planning_scene.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_ros/moveit_cpp/planning_component.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_ros/moveit_cpp/moveit_cpp.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.hpp
/root/ws_moveit/src/moveit2/moveit_py/src/moveit/moveit_ros/planning_scene_monitor/planning_scene_monitor.hpp
/root/ws_moveit/src/moveit2/moveit_core/constraint_samplers/test/pr2_arm_kinematics_plugin.hpp
/root/ws_moveit/src/moveit2/moveit_core/constraint_samplers/test/pr2_arm_ik.hpp
/root/ws_moveit/src/moveit2/moveit_ros/moveit_servo/tests/servo_ros_fixture.hpp
/root/ws_moveit/src/moveit2/moveit_ros/moveit_servo/tests/servo_cpp_fixture.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/cartesian_path_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/state_validation_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/move_action_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/execute_trajectory_action_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/tf_publisher_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/apply_planning_scene_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/kinematics_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/save_geometry_to_file_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/load_geometry_from_file_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/clear_octomap_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/move_group/src/default_capabilities/plan_service_capability.hpp
/root/ws_moveit/src/moveit2/moveit_ros/planning/trajectory_execution_manager/test/test_moveit_controller_manager.hpp
```

All of these are due to no `include` directory. I'm unsure how to handle these cases (in the script, or manually) so some guidance would be appreciated here.

**Comment by TSNoble** [comment]
I just noticed the "autogenerated by script" lines should likely be comments

**Comment by TSNoble** [comment]
The tutorial images seem to be passing, but all I did was comment out the `#pragma message`, which shouldn't raise a warning.

@sea-bass @sjar

I'm a little concerned about the files mentioned above missing include directories. These mainly seem to be tests, `moveit_py`, and `default_capabilities`. I have a few questions:

  - Does this imply they are only being imported locally, and don't form part of the public API?
  - If so, should files with no include parent directory be ignored by the auto-gen script, rather than raising a warning?
  - If not (or perhaps regardless), should these be modified to follow the same directory structure: 
     - ìncliude/moveit/...`, `src/moveit/...`
     - Is there any reason they deviate from the norm, besides for convenience for importing locally?

**Comment by TSNoble** [comment]
The humble failure is also appearing on master, so unrelated: https://github.com/moveit/moveit2/actions/runs/11966505405/job/33362233571

I suppose this could just be a follow-on failure since @sea-bass fixed the PlanningScene rests?

**Comment by TSNoble** [comment]
@sjahr @sea-bass

I believe this is ready for review now. Summary, since the thread is quite long:

- I've replaced all .h files with .hpp (except ikfast and the autogenerated exports)
- I've added a script which autogenerates .h files from .hpp files (with BSD-3, Authoriship, and non-warning deprecation msg)
- I've run this manually over the source tree and committed the .h files
- The only files I wasn't able to generate .h files for were ones that aren't being installed (tests, default_capabilities, moveit_py)
- Hopefully, the passing CI pipelines show that the imports work internally
- Hopefully, the passing Tutorial pipelines show that the imports work externally
- I'd think I'd like to include auto-gen of headers in the CI, but this can wait until we've seen the method works well
- Failing humble build seems to be unrelated

**Comment by rr-tom-noble** [comment]
@sea-bass 

Would you mind running the humble-ci job on this PR again when https://github.com/moveit/moveit2/issues/3130 is resolved, just to make sure everything passes?

File: moveit/scripts/create_deprecated_headers.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Copyright 2021 PickNik Inc.
```
Comment (sea-bass) [inline moveit/scripts/create_deprecated_headers.py:4]: 2024, and you should feel free to either use your own name or your company's here.

File: moveit/scripts/create_deprecated_headers.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+class NoIncludeGuard(Exception):
+    ERROR = "No include guard found in {}.hpp. Unable to generate pretext."
+
+    def __init__(self, file: Path):
+        super().__init__(self.ERROR.format(file))
+
+
+class NoIncludeDirectory(Exception):
+    ERROR = "No includue directory found for {}.hpp. Unable to generate relative .hpp include"
```
Comment (sea-bass) [inline moveit/scripts/create_deprecated_headers.py:48]:
```suggestion
    ERROR = "No include directory found for {}.hpp. Unable to generate relative .hpp include"

**Review by sea-bass — COMMENTED** [review]
CI is all green -- gave your script a look, and obviously not the headers :)

File: moveit/scripts/create_deprecated_headers.py (hunk 1 above)
Comment (rr-tom-noble) [reply moveit/scripts/create_deprecated_headers.py:4]: Oh awesome :smile:

File: moveit/scripts/create_deprecated_headers.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    apply = "--apply" == sys.argv[1] if n_args == 2 else False
+    return apply
+
+
+if __name__ == "__main__":
+    apply = parse_args(sys.argv)
+    hpp_paths = list(Path.cwd().rglob("*.hpp"))
+    print(f"Generating deprecated .h files for {len(hpp_paths)} .hpp files...")
+    processed = []
+    bad = []
```
[the remark this answers is not in the document]
Comment (rr-tom-noble) [reply moveit/scripts/create_deprecated_headers.py:110]: Will do :+1:

File: moveit/scripts/create_deprecated_headers.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        items = [self.prefix, self.hpp.pretext, self.warn, self.hpp.include]
+        return "\n\n".join(items)
+
+
+def parse_args(args: List) -> bool:
+    n_args = len(sys.argv)
+    if n_args > 2:
+        sys.exit("Usage: ./create_deprecated_headers [--apply]")
+    apply = "--apply" == sys.argv[1] if n_args == 2 else False
+    return apply
```
[the remark this answers is not in the document]
Comment (rr-tom-noble) [reply moveit/scripts/create_deprecated_headers.py:102]: Haha yeah I was thinking the same. Will do :+1:

File: moveit/scripts/create_deprecated_headers.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        relative_import = self.path.relative_to(include_paths[0])
+        return f"#include <{relative_import}>"
+
+
+class DeprecatedHeader:
+    def __init__(self, hpp: HppFile):
+        self.hpp = hpp
+        self.path = hpp.path.with_suffix(".h")
+        self.prefix = f"/* This file was autogenerated by {Path(__file__).name} */"
+        self.warn = '#pragma message(".h header is obsolete. Please use the .hpp")'
```
[the remark this answers is not in the document]
Comment (rr-tom-noble) [reply moveit/scripts/create_deprecated_headers.py:89]: That's fine. I'll likely make all of these changes this evening (once I'm off the clock)

File: moveit/scripts/create_deprecated_headers.py (hunk 1 above)
Comment (rr-tom-noble) [reply moveit/scripts/create_deprecated_headers.py]: Done :+1:

File: moveit/scripts/create_deprecated_headers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TSNoble) [reply moveit/scripts/create_deprecated_headers.py]: Done :+1:

File: moveit/scripts/create_deprecated_headers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TSNoble) [reply moveit/scripts/create_deprecated_headers.py]: Done (along with making some changes so the .h files satisfy clang-format)👍

**Comment by tylerjw** [comment]
You should merge this so my email inbox can rest again :laughing:

**Comment by rr-tom-noble** [comment]
@sea-bass

I've applied your suggestions now (and tidied up the script a little more, and made the output a little nicer). I'd also be in favour of merging this soon so we can deal with any merge conflicts with other PRs quickly, but I also understand if you want to wait until the maintainer meeting tomorrow

**Review by nbbrooks — COMMENTED** [review]
I suggest adding a bit more context to the autogenerated message at the top of the h files. Your changelog notes are very descriptive, how about we use the:

`/* .h This file was autogenerated by create_deprecated_headers.py */`
becomes
`/* .h All MoveIt 2 headers have been updated to use the .hpp extension. .h headers are now autogenerated with a deprecation warning, and may be removed in future releases. Please update your imports to use the .hpp headers. */`

**Comment by TSNoble** [comment]
> I suggest adding a bit more context to the autogenerated message at the top of the h files. Your changelog notes are very descriptive, how about we use the:
> 
> `/* This file was autogenerated by create_deprecated_headers.py */` becomes `/* All MoveIt 2 headers have been updated to use the .hpp extension. .h headers are now autogenerated with a deprecation warning, and may be removed in future releases. Please update your imports to use the .hpp headers. See https://github.com/moveit/moveit2/pull/3113 for details. */`

@nbbrooks

Thanks for the suggestion. I've modified the script (with a few tweaks to the formatting for readability & consistency with the BSD-3 comments), and have regenerated the header files.

**Review by sea-bass — APPROVED** [review]
Let's do this! Thanks again @TSNoble

**Comment by rr-tom-noble** [comment]
@sea-bass 

Humble tests failed on the merge queue with what looks like https://github.com/moveit/moveit2/issues/3130 :cry:

**Comment by sea-bass** [comment]
Ah yes I need to merge the other PR first. One sec

**Comment by sea-bass** [comment]
I'm gonna go ahead and merge this since the deprecation warning in https://github.com/moveit/moveit2/pull/3139 just snuck in today and I want to keep things moving for now.

**Pull request merged**